crates/core/src/replication/connection.rs

```diff
@@ -204,19 +253,58 @@ impl RemoteStatement {
     pub async fn prepare(conn: RemoteConnection, sql: &str) -> Result<Self> {
         let stmts = parser::Statement::parse(sql).collect::<Result<Vec<_>>>()?;
 
-        let local_statement = if conn.should_execute_local(&stmts[..]) {
-            let stmt = conn.local.prepare(sql).await?;
-            Some(stmt)
-        } else {
-            None
-        };
-
+        // FIXME(sarna): this condition was originally there, but should it? We're in RemoteStatement,
+        // so it's kind of assumed now we're not executing locally
+        //
+        // if conn.should_execute_local(&stmts[..]) {
+        //     println!("Preparing {sql} locally");
+        //     let stmt = conn.local.prepare(sql).await?;
+        //     return Ok(Self {
+        //         conn,
+        //         stmts,
+        //         local_statement: Some(stmt),
+        //         metas: vec![]
+        //     })
+        // }
```

Comment:
So now we never run any prepared statements locally? That was the goal of the original if statement was to parse the entire sql statement to see if there were any writes. This + connection state determined if it was remote or local. I think maybe what we want is a enum now that is

```rust
enum RemoteStatementState {
    Local(LibsqlStatement),
    Remote(Vec<StatementMeta>)
}
```

and based on this we will execute the statement locally or remote.

Reply:
The issue I hit was that `SELECT 1` from here: https://github.com/aqrln/libsql-tx-repro/blob/82ab5c5ee7386b0650244ad33ee574e86e10acb5/test.mjs#L17-L18
got prepared locally even though it was already part of a write transaction. And after it got prepared locally, we also seemed to have executed it locally, and the whole example crashed because "execute returned rows". I didn't exactly have cycles to debug the javascript part, but I figured that if we end up in `RemoteStatement`, it already means we ruled out running the statement locally, right?

If the intention here was to optimize readonly prepare(), it won't work in general case, because prepare() in SQLite takes the db state into account. For instance, if you forward `CREATE TABLE t(id)` to the primary, and then try to prepare `SELECT * FROM t` locally, it will fail with a "no such table" error, even though the query is legal and read-only.

Reply:
Wouldn't we just want to reroute to the primary if we hit that case?